### ISSUE TITLE
#[question] why ppo multiprocess results are worse than single process ?

### ISSUE THREAD

**Issue by greg2paris** [hill-a/stable-baselines#911]
Hello,
I trained my agent on single environment on my laptop and i got better results than when i trained it on google TPU in Colab.

When I train in google colab using TPU i am using 40 environment.
I use this line to know how many cpu i can use :
`number_of_processors = multiprocessing.cpu_count()`

I am using ppo with this settings:
policy_type = "MlpPolicy"
learning_rate = 0.00001  # * n_cpu
gamma = 0.99 
n_steps = 128  
ent_coef = 0.01
nminibatches = 4  # * n_cpu
noptepochs = 4 
cliprange = 0.2 
cliprange_vf = None

I dont get why I am getting worse results, do i have to scale the learning rate and the minibatch sise with the number of processor?
What did I do wrong?

**Comment by Miffyli** [comment]
All those settings are irrelevant of number of CPUs on the machine. Did you also use 40 environments on laptop? More environments -> more even sampling of environment experiences all around -> stabler learning -> (likely) better results. Also results can vary significantly per run, so you need to run multiple training runs for fair comparison. See [docs](https://stable-baselines.readthedocs.io/en/master/guide/rl_tips.html).

**Comment by greg2paris** [comment]
The way

> All those settings are irrelevant of number of CPUs on the machine. Did you also use 40 environments on laptop? More environments -> more even sampling of environment experiences all around -> stabler learning -> (likely) better results. Also results can vary significantly per run, so you need to run multiple training runs for fair comparison. Also see [docs](https://stable-baselines.readthedocs.io/en/master/guide/rl_tips.html).

Correct me if i am wrong, but the way I understand it is that:
The batch size is equal to the n_steps multiplied by the number of environment.
Then the mini-batch size is equal to the batch size divided by the ppo nminibatches value.
It means that when I train on a single environment the batch size would be 128 (128 * 1).
And the mini-batch size would be 32 (128 / 4).

But when I train on 40 cpu, the batch size should be 5120 (128 *40).
And the mini-batch size would be 1280 (5120 / 4).

So maybe the mini-batch size has to be adapted to the number of environment?

**Comment by Miffyli** [comment]
> So maybe the mini-batch size has to be adapted to the number of environment?

Yes, this is correct, but number of environments is _not_ tied to number of cores. You can launch 40 environments even on a single-core machine and should get similar results, just a bit slower. Generally, more environments is better learner, even though it might not be faster steps-per-second.

**Comment by greg2paris** [comment]
> > So maybe the mini-batch size has to be adapted to the number of environment?
> 
> Yes, this is correct, but number of environments is _not_ tied to number of cores. You can launch 40 environments even on a single-core machine and should get similar results, just a bit slower. Generally, more environments is better learner, even though it might not be faster steps-per-second.

When I launch it on my laptop I get 200 fps on average, and when It in colab I have around 4000fps. But the results are not as good.
So I was wondering if we have to adjust the learning rate to the big mini-size 1280, or should we modify the mini-batch value so i get the mini-batch size of 32.

**Comment by Miffyli** [comment]
You should not need to adjust hyperparameters to get same results. You should get similar results if:
- Python and library versions match, along with underlying math libraries etc
- Hyperparameters are same
- Number of environments match
- Things are set [deterministic](https://stable-baselines.readthedocs.io/en/master/guide/algos.html#reproducibility)

In short, you should not see different results on different machines if above are same, even if you have less CPUs on other machine.

**Comment by araffin** [comment]
i recommend you reading the multiprocessing part if our tutorial: https://github.com/araffin/rl-tutorial-jnrr19
in short, multiprocessing means usually faster training time but less sample efficient.

**Comment by greg2paris** [comment]
Thank you for the answer, their is lot of things to learn their
[https://colab.research.google.com/github/araffin/rl-tutorial-jnrr19/blob/master/3_multiprocessing.ipynb#scrollTo=Mu8ZYFQN3yLr]
